### ISSUE TITLE
[Feature Request] Virtual File System with on-demand file loading

### ISSUE THREAD

**Issue by YanzheL** [meganz/MEGAsync#430]
## Background
MEGAsync is a sync client, which means the local and cloud storage should match.
If I want to sync 2T files in cloud, then I need 2T storage in local, which is not convenient.

Compared to Google Drive, they have two types of client app:
`Backup & Sync`, and `Google File Stream`. Currently, MEGAsync is an alternative to `Google Backup & Sync`, but not `Google File Stream`.

## Expected to see

I think we should have a new app similar to `Google File Stream`, which mounts a virtual disk via FUSE driver. All files on the cloud are available in that virtual drive, and files are downloaded only if user open it or choose `available offline`. Incremental uploading is triggered only if the file is modified.
Microsoft OneDrive have similar feature too.

The advantage of this feature is obvious, I just don't understand why there is no such app for MEGA. This is a main blocker for me to migrate from Google Drive.
The new app can completely replace MEGAsync.

**Comment by themagiulio** [comment]
Hey!
I found a way to stream the whole MEGA Cloud Drive using MEGAcmd.
I managed to do that using [webdav](https://github.com/meganz/MEGAcmd/blob/master/UserGuide.md#webdav).

So you need to:
1) Download and execute [MEGAcmd](https://mega.nz/cmd).
2) Login with you account.
3) Type `webdab ./` to stream the whole drive (alternatively you can stream every path you want).
4) The app will prompt an url which you can add to Explorer as a network resource.
5) And you're done, you can stream, edit, add files.

NOTE: As it is said in the User Guide, webdav is an experimental feature, so it may have some problems.

I hope I helped you.

**Comment by NicoHambauer** [comment]
Thanks for opening this issue! I im also looking forward to the implementation of your feature request! With iCloud Drive on windows for example one can decide to safe storage and check or uncheck files, folders or even everythin so the sync app doesnt use any storage on your computer, but you see all items! (I could post a screenshot here, but it would only show like: size of your uploads: 40GB, size on disk: 10MB) This would be very useful for mega.nz/MegaSync as well! If this feature would be implemented i would also go for the 2-8TB Version!!

**Comment by pwnorbitals** [comment]
+1 and issue bump

**Comment by timh-86** [comment]
We definitely need this feature. 
I want to move away from Google Drive because it's buggy and well... it's Google.

But I need virtual drive for at least part of my cloud storage. 
If Mega does not deliver I'll reluctantly move to pCloud. ;(

**Comment by paulverbeke** [comment]
With the recent research paper regarding cryptography implementation problems on the MEGA platform (https://arstechnica.com/information-technology/2022/06/mega-says-it-cant-decrypt-your-files-new-poc-exploit-shows-otherwise) I'd say they are currently working on something else 😅. And this feature is definitely not for 2022 🥲. I'll wager that Proton Drive will ship it's desktop app before this MEGA feature

**Comment by ZPNRG** [comment]
Yes, MEGA needs something like the OneDrive Files On-Demand or Dropbox Smart Sync feature where the user can see all the files that are in the MEGA cloud, but choose whether or not they are synced down to the PC (i.e. actually available locally / offline). 

I haven't used Dropbox Smart Sync, but overall OneDrive Files On-Demand works quite well.  I like the three (3) statuses Microsoft has implemented where files/folders are... 

1) "Online-only files" (only stored in the cloud), 

2) files are temporarily downloaded and available locally (offline), but then possibly deleted locally and made again "Online-only files", all depending on the user's Storage Sense settings, or 

3) "Always available files" that are synced locally and in the cloud and remain locally available unless the user executes the "Free up space" option to make that file/folder "Online-only files".


MEGA could keep it simple and have only OneDrive's #1 and #3 status for files.  They are either  "Online-only files" or "Always available files", but either way, you can see all of the files/folders you have stored in the MEGA cloud whether or not they are synced locally and already downloaded.

I primarily use Windows, but this type of feature would be just as important for Mac users and I do sometimes use a Mac and I have family and friends that do too.  In my experience of using MEGA, this missing capability is the biggest negative compared to OneDrive and Dropbox. 

As a side note:  MEGA also needs to join OneDrive and Dropbox in using "Block-Level Sync" so that files are split up into smaller pieces and if/when a change is made to a file, only the part that changed gets synced, removing the need to sync the whole file again. This is especially important with large files.

See https://support.microsoft.com/en-us/office/save-disk-space-with-onedrive-files-on-demand-for-windows-0e6860d3-d9f3-4971-b321-7092438fb38e for more details.

**Comment by boretom** [comment]
I'm a long time user of MEGA (free version though) and that requirement has now come up.

Wanting to avoid the other mentioned services it would be great to get that feature. Maybe we could get some words from @mattw-mega or someone else from MEGA if there's a chance that get's on the roadmap?

**Comment by paulverbeke** [comment]
Guess what ? I was right: Proton Drive shipped its Windows app in stable and macOS app in beta, both with on-demand sync, cached files, and auto free space after a set time.
I don't see the Mega sync app with on-demand sync 🕵️‍♂️

**Comment by dentropy** [comment]
Thanks guys for this thread it answered my questions nicely, I expect it would be pretty simple to write a fuse file system using the mega CLI tools, might not be real time sync though. Also I wonder if rclone is relevant to this

**Comment by VictorQueiroz** [comment]
> With the recent research paper regarding cryptography implementation problems on the MEGA platform (https://arstechnica.com/information-technology/2022/06/mega-says-it-cant-decrypt-your-files-new-poc-exploit-shows-otherwise) I'd say they are currently working on something else 😅. And this feature is definitely not for 2022 🥲. I'll wager that Proton Drive will ship it's desktop app before this MEGA feature

It is also not for 2024. I was thinking of getting a subscription, but it seems so troublesome to achieve this. I'll wait for it to be available.

**Comment by ZPNRG** [comment]
I think I am going to go ahead and seriously try Sync.com.  They have CloudFiles which is perhaps the best alternative to OneDrive Files On-Demand or Dropbox Smart Sync.  Proton also offers something similar on Windows, though they are using Microsoft's OneDrive APIs, etc. to provide that.

**Comment by rautamiekka** [comment]
> Sync.com.

Not an option cuz no Linux with a quick check of the download page.

**Comment by rgheno** [comment]
I would love to give Mega a shot, specially because it charges by storage, not by users. It's a big problem that when I have a new employee I have to pay an eye for a new google user and it comes with more TBs. The same with dropbox, if I have 4 users with 3TB and I add another user I will get, lets say, another TB, but another employee definitely wont demand a whole nother TB, if you know what I mean. Mega was the first one I found that charged by GB and behaved like a normal drive.

**Comment by RobATX** [comment]
It would be nice to hear from Mega on this issue. This is something that prevents the service from truly competing with iCloud, OneDrive, and Dropbox.

**Comment by vtmateos** [comment]
Dear Users,

Thank you for your feedback.

We are currently developing a short-term solution to integrate FUSE capabilities into Windows systems, tailored to meet your specific needs. In the long term, we plan to implement a Windows Cloud API solution on top of our sync engine, ensuring seamless integration with Windows.

Stay tuned for future updates.

Regards.

**Comment by schreibfheler** [comment]
What about support on mac?

**Comment by ialexpw** [comment]
Purchased a subscription before reading this (as I wrongly assumed it would already be there) - and now had to cancel it again, nearly every other cloud service has this feature already.

**Comment by Humblean** [comment]
Hey Mega Team, please tell us you're working on this! Seems also like it's quite trvial to implement.

**Comment by ray8828** [comment]
what the fuck is this progress? 5 years???????? fuck mega

**Comment by ZPNRG** [comment]
> what the fuck is this progress? 5 years???????? fuck mega

Take your ignorant trash talking comment somewhere else. I agree that MEGA has been far too slow and maybe even incompetent in addressing this issue, but just say that.

**Comment by ZPNRG** [comment]
> We are currently developing a short-term solution to integrate FUS capabilities into Windows systems, tailored to meet your specific needs. In the long term, we plan to implement a Windows Cloud API solution on top of our sync engine, ensuring seamless integration with Windows.
> 
> Stay tuned for future updates.

@vtmateos 

1) Status update?

2) It might be better to just go straight to the long-term and/or best solution. MEGA has already taken far too long with this. Proton (Proton Drive) came straight out of the gate with support for “files on demand” using Microsoft’s Cloud API to semi-mimic OneDrive support.

3) The Proton Drive Windows client is horribly slow to sync. Obviously, that would be a huge downgrade if MEGA ended up with that problem.

**Comment by fcnjd** [comment]
>     2. It might be better to just go straight to the long-term and/or best solution. MEGA has already taken far too long with this. Proton (Proton Drive) came straight out of the gate with support for “files on demand” using Microsoft’s Cloud API to semi-mimic OneDrive support.
> 
I totally agree on that. Additionally, not only Proton got that done, now all the major open source cloud software platforms (namely Nextcloud, Owncloud and even Seafile) have integrated the Windows cloud files sync API in their desktop clients, leaving the MEGAsync client behind. I think we can say that virtual files has become a requirement rather than a feature, as users are expecting it as normal.

**Comment by nohde** [comment]
@vtmateos please any useful updates for this? Highly requested feature and I think one of the biggest reason a lot people can’t/won’t switch. I love mega but I/we need this really basic feature. This issue is now 5.5 year here and no native solution by now. Thank you…

**Comment by vtmateos** [comment]
Hi there,

This project remains a key priority for us, however, we are currently finalizing a few other priority projects.

Apologies for the delay. We will provide an update as soon as we reach the next major milestones.

Kind regards,

**Comment by nohde** [comment]
@vtmateos 

Hi, if I am allowed to ask, what are the other priority projects which were the reason to hold back a industry standard feature for more than 6 years ? Is there any roadmap with more details or where we can see which feature is planned when ? How many months away are the current major milestones ?

**Comment by EmreCanOezcan** [comment]
I just switched from Dropbox thinking that files on demand would of course already be implemented by now, but as I was using rclone to copy over my files, I noticed that I don't have the ability to choose to store files only in the cloud.
I was really excited to switch to MEGA but am disappointed on day one.

I understand that software development may not go as planned and that roadblocks are always a thing, but 6 years for this implementation without any real roadmap and lack of transparent communication is just sad! 👎

**Comment by timh-86** [comment]
Maybe trie Pcloud instead. They DO offer selective syncing (local / cloud) I think.

Pricing is better than Dropbox & it's a EU, privacy friendly company.



________________________________
Van: Felt ***@***.***>
Verstuurd: dinsdag, januari 27, 2026 11:08:36 p.m.
Aan: meganz/MEGAsync ***@***.***>
CC: timh-86 ***@***.***>; Comment ***@***.***>
Onderwerp: Re: [meganz/MEGAsync] [Feature Request] Virtual File System with on-demand file loading (#430)

[https://avatars.githubusercontent.com/u/61991850?s=20&v=4]EmreCanOezcan left a comment (meganz/MEGAsync#430)<https://github.com/meganz/MEGAsync/issues/430#issuecomment-3807770307>

I just switched from Dropbox thinking that files on demand would of course already be implemented by now, but as I was using rclone to copy over my files, I noticed that I don't have the ability to choose to store files only in the cloud.
I was really excited to switch to MEGA but am disappointed on day one.

I understand that software development may not go as planned and that roadblocks are always a thing, but 6 years for this implementation without any real roadmap and lack of transparent communication is just sad! 👎

**Comment by ZPNRG** [comment]
> I understand that software development may not go as planned and that roadblocks are always a thing, but 6 years for this implementation without any real roadmap and lack of transparent communication is just sad! 👎

I don't think there is any way to sugarcoat the fact that the MEGA devs, management, and/or leadership team either 1) do not know what they are doing and/or 2) they are not taking this request very seriously and are really just giving "lip service" to customers.

I and others in the past have also requested that they join OneDrive, Dropbox, and others in using "Block-Level Sync" so that files are split up into smaller pieces. If/when a change is made to a file, only the part that changed gets synced, removing the need to sync the whole file again. This is especially important with large files. There have been questions about doing this with encryption, but my understanding is it can be done.

We have seen no reason to believe that there is any legitimate attempt to address these issues by MEGA, just a lot of talk. Most of the changes (especially features) in updates being made to the MEGAsync client do not seem to be that critical.

**Comment by rautamiekka** [comment]
> > I understand that software development may not go as planned and that roadblocks are always a thing, but 6 years for this implementation without any real roadmap and lack of transparent communication is just sad! 👎
> 
> I don't think there is any way to sugarcoat the fact that the MEGA devs, management, and/or leadership team either 1) do not know what they are doing and/or 2) they are not taking this request very seriously and are really just giving "lip service" to customers.
> 
> I and others in the past have also requested that they join OneDrive, Dropbox, and others in using "Block-Level Sync" so that files are split up into smaller pieces. If/when a change is made to a file, only the part that changed gets synced, removing the need to sync the whole file again. This is especially important with large files. There have been questions about doing this with encryption, but my understanding is it can be done.
> 
> We have seen no reason to believe that there is any legitimate attempt to address these issues by MEGA, just a lot of talk. Most of the changes (especially features) in updates being made to the MEGAsync client do not seem to be that critical.

Chill the fuck out already you dickheads, you're only worsening things.

Yes, 1 has truth to it as we've seen (they went all in into things they had no experience in, but they've proven they wanna keep learning. When you're learning on the go, no matter how much staff you have, it takes time). 2, no, that's bullshit.

This filesystem thing is very complicated, too; you simply can't make such a thing easily, especially when trying to keep it cross-platform across 3 very different ones, of which 1 is an order of magnitude more complicated due to the community fragmentation.

**Comment by ZPNRG** [comment]
@rautamiekka, watch your language. There is no need or place for that here.

I've been a computer scientist / engineer for going on 25 years. I didn't say anything was "easy".  Look, this is business. Sorry. If you have been a paying customer for years as I have been, then waiting for certain things for a year or two is one thing. Six years. It is ridiculous. Cry me a river about it being hard or complicated. Competition is hard.

I pay for and use other solutions too. I am evaluating other options and have been. MEGA still works fairly well as a whole despite my complaints. If not, I would have already stopped paying for MEGA and dropped it.

**Comment by ialexpw** [comment]
> Chill the fuck out already you dickheads, you're only worsening things.

@rautamiekka and you think saying that is OK?

There are existing libraries that other companies use to integrate this depending on the OS. Basically every other cloud service has this feature and it is not as complicated as you "believe" it is. Dropbox, Sync, Proton, pCloud, Filen (of which had 1 software developer when this was implemented) etc etc.

Going on 6 years now is ridiculous and they're losing customers due to this.

**Comment by ZPNRG** [comment]
> There are existing libraries that other companies use to integrate this depending on the OS. Basically every other cloud service has this feature and it is not as complicated as you "believe" it is. Dropbox, Sync, Proton, pCloud, Filen (of which had 1 software developer when this was implemented) etc etc.

Correct. I knew Dropbox and Proton Drive were using Microsoft's Windows Cloud Files API, at least on Windows, though maybe they are for the macOS client too. Dropbox didn't use that API, but switched to it and you could clearly see the change in the Windows client a number of months ago. I wasn't sure about pCloud (haven't really looked at that in many years) and I've never closely checked out Filen. It is nice to know they have the functionality.

**Comment by KOXXPL** [comment]
Any updates on this feature?

**Comment by uneydev** [comment]
I was finally told from support a few days ago, that in early july the closed beta will be published. Full file sync so no block sync but it will be released for all clients Windows, Mac and Linux at the same time
